src/Engine/ProtoCore/DSASM/Executive.cs

```diff
@@ -6962,14 +6982,21 @@ private void GC()
 #if DEBUG
                 gcRootSymbolNames.Add("__thisptr");
 #endif
-                gcRoots.Add(stackFrame.ThisPtr);
+                if (stackFrame.ThisPtr.IsReferenceType)
+                    gcRoots.Add(stackFrame.ThisPtr);
                 blockId = stackFrame.FunctionCallerBlock;
                 currentFramePointer = stackFrame.FramePointer;
             }
+#endif
+            if (RX.IsReferenceType)
+                gcRoots.Add(RX);
 
-            gcRoots.Add(RX);
+#if TRACING_GC
+            gcRoots.AddRange(runtimeCore.ReplicationRoots.Where(r => r.IsReferenceType));
```

Comment:
This is a fairly inefficient style of coding that LINQ encourages. Is this a hot path? Should we look at storing them in a separate list?